tests/kafkatest/services/kafka/kafka.py

```diff
@@ -593,6 +593,9 @@ def close_port(self, listener_name):
 
     def start_minikdc_if_necessary(self, add_principals=""):
         has_sasl = self.security_config.has_sasl
+        # Since KafkaService is utilized by both controller and broker, we do not set miniKDC to None.
+        # This avoids the creation of an additional miniKDC, which could result in a mismatch
+        # between client and server tokens if two miniKDC instances are running concurrently.
         if has_sasl:
```

Comment:
One question: this implies that has_sasl returns `false` for one of the controller or broker?

Reply:
This depend on `KafkaService` config.
When controller or broker is created, it will check their configs and decide to wherther utilize SASL.